llvm/include/llvm/CodeGen/TargetRegisterInfo.h

```diff
@@ -42,11 +42,27 @@ class VirtRegMap;
 class LiveIntervals;
 class LiveInterval;
 
+/// TargetSuperClassIterator enumerates all super-registers of RegClass.
+class TargetSuperClassIterator
+    : public iterator_adaptor_base<TargetSuperClassIterator, const unsigned *> {
+public:
+  /// Constructs an end iterator.
+  TargetSuperClassIterator() = default;
+
+  TargetSuperClassIterator(const unsigned *V) { I = V; }
+
+  const unsigned &operator*() const { return *I; }
```

Comment:
It should return `const TargetRegisterClass &`

Reply:
Ah I see it won't always work, some users need the ID, not TargetRegisterClass.
Never mind.

Reply:
That's not an obstacle. The bigger problem is that you need a TargetRegisterInfo to do the mapping from IDs back to TargetRegisterClass's, which we can't easily get from inside the iterator.

Reply:
One way to do it is to move `superregclasses` to `TargetRegisterInfo` and pass the iterated class as an argument. (This is how similar `MCRegisterInfo` methods are implemented.)
Alternatively, pass `const TargetRegisterInfo &` to `superclasses` and then into the iterator constructor. Not sure if this is worse or better.

That is, it is possible, but since the main goal of this PR is to reduce the number of relocations I think it could be done separately.